### PR TITLE
kazana / app / plugin namespaces

### LINKED ISSUE #17: Improve plugin config strategy
Improve config of plugins so that the config items are accessed in the same way regardless of if the plugin is started as Hapi plugin or as a standalone app.

### PR DESCRIPTION
**Pull request by gr2m**
### Problem with plugin configuration before

If I have this `.kazanarc`

```ini
[admin]
  username = 'myadmin'
```

And my app’s `index.js` is

```js
module.exports = {
  name: 'myApp',
  config: {
    db: {
      rawData: 'my-raw-data',
    },
    mySetting: 'value'
  },
  plugins: [{
    name: 'myPlugin',
    config: {
      foo: 'baz'
    }
  }]
  }
}
```

Then my config gets set to

```js
server.setting.app.admin.username === 'myadmin'
server.setting.app.db.rawData === 'my-raw-data'
server.setting.app.mySetting === 'value'
server.setting.app.plugin.myPlugin.foo === 'baz'
```

Problem is that when "myPlugin" is run in a separate process,
its config gets set to

```js
server.setting.app.foo === 'baz'
```

### Suggested solution

To resolve the issue with the configuration being at different places,
I think the cleanest way would be to introduce a `kazana` namespace
for the default settings like admin & database names.

The `.kazanarc` would be changed to

```ini
[kazana]
  [admin]
    username = 'myadmin'
```

The `index.js` from the example above would be changed to

```js
module.exports = {
  name: 'myApp',
  config: {
    kazana: {
      db: {
        rawData: 'my-raw-data',
      }
    },
    myApp: {
      mySetting: 'value'
    }
  },
  plugins: [{
    name: 'myPlugin',
    config: {
      foo: 'baz'
    }
  }]
  }
}
```

Now that we have a kazana config namespace, we can use the app & plugin
names as namespaces to

```js
server.setting.app.kazana.admin.username === 'myadmin'
server.setting.app.kazana.db.rawData === 'my-raw-data'
server.setting.app.myApp.mySetting === 'value'
server.setting.app.myPlugin.foo === 'baz'
```

That way it wouldn’t matter if a plugin gets loaded standalone or part
as the main process, the configuartion would always be set accessible
at `server.setting.app.myPlugin`

With that setting, I could overwrite the default settings from the plugin
in my app like this

```js
module.exports = {
  name: 'myApp',
  config: {
    kazana: {
      db: {
        rawData: 'my-raw-data',
      }
    },
    myApp: {
      mySetting: 'value'
    },
    myPlugin: {
      foo: 'new baz!'
    }
  },
  plugins: [{
    name: 'myPlugin',
    config: {
      foo: 'baz'
    }
  }]
  }
}
```

### How to not break all the things

All configuration keys must remain accessible at `server.setting.app` as they
are today. So if the namespaces are used, we will set both.

If namespaces are not used, we will show deprecation warnings. That should™ be it.

closes #17

### CODE REVIEW

**Comment by gr2m** [comment]
I’ve tested the existing example app and van with the new config, without any farther changes, and it worked just fine, besides the deprecation warnings being shown.

File: tests/unit/plugins-test.js [hunk 1]
```diff
@@ -22,11 +26,11 @@ function mockRc (name, config) {
 }
 getConfig.rc = mockRc
 
-test('app config', function (t) {
-  t.is(getConfig(kazanaApp).foo, 'bar', 'parses app.config')
-  t.is(rcName, 'kazana', 'passes "kazana" as name to rc')
-  t.end()
-})
+// test('app config', function (t) {
```
Comment (jo) [inline tests/unit/plugins-test.js]: Wouldn't be `test.skip` be better than commenting out?

**Comment by jo** [comment]
@gr2m great work! I very much like the idea of prefixing kazana internal config.

Since this is a quiet complex pr I have some questions, please bear with me :baby_chick:

**Comment by gr2m** [comment]
Great feedback, thanks!

File: tests/unit/plugins-test.js (hunk 1 above)
Comment (gr2m) [inline tests/unit/plugins-test.js]: ooops, only commented it out temporarily, thanks :)

**Comment by gr2m** [comment]
@jo I've pushed another commit. I decided to add proper integration tests, and that was a good idea, I found several issues that I missed before. 

But now I'm quite confident that we won’t break anything. And if we do, we should be able to very quickly add an integration test to reproduce the problem, and then solve it.

If you think the changes look good, feel free to clean up the commits and merge

**Comment by jo** [comment]
This is how tests should look like:
```
tests/integration/integration-test.js ............... 13/13
tests/unit/defaults-test.js ......................... 16/16
tests/unit/plugins-test.js ............................ 4/4
total ............................................... 33/33

  33 passing (3s)

  ok
---------------------------|----------|----------|----------|----------|----------------|
File                       |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
---------------------------|----------|----------|----------|----------|----------------|
 __root__/                 |      100 |      100 |      100 |      100 |                |
  index.js                 |      100 |      100 |      100 |      100 |                |
 lib/                      |      100 |      100 |      100 |      100 |                |
  defaults.js              |      100 |      100 |      100 |      100 |                |
  get-plugin-defaults.js   |      100 |      100 |      100 |      100 |                |
  parse-manifest-config.js |      100 |      100 |      100 |      100 |                |
  parse-rc-config.js       |      100 |      100 |      100 |      100 |                |
 tests/integration/        |      100 |      100 |      100 |      100 |                |
  run-example.js           |      100 |      100 |      100 |      100 |                |
---------------------------|----------|----------|----------|----------|----------------|
All files                  |      100 |      100 |      100 |      100 |                |
---------------------------|----------|----------|----------|----------|----------------|
```

Great work @gr2m :+1:

**Pull request merged**